x/stakeibc/keeper/msg_server_submit_tx.go

```diff
@@ -165,6 +167,143 @@ func (k Keeper) ReinvestRewards(ctx sdk.Context, hostZone types.HostZone) error
+	return nil
+}
+
+// icq to read host delegation account undeleted balance => update hostZone.delegationAccount.Balance
+// TODO(TEST-97) add safety logic to query at specific block height (same as query height for delegated balances)
+func (k Keeper) UpdateUndelegatedBalance(ctx sdk.Context, index int64, hostZone types.HostZone) error {
+	_ = ctx
+	// Fetch the relevant ICA
+	delegationAccount := hostZone.GetDelegationAccount()
+
+	var cb icqkeeper.Callback = func(icqk icqkeeper.Keeper, ctx sdk.Context, args []byte, query icqtypes.Query) error {
+
+		queryRes := bankTypes.QueryAllBalancesResponse{}
+		err := k.cdc.Unmarshal(args, &queryRes)
+		if err != nil {
+			k.Logger(ctx).Error("Unable to unmarshal balances info for zone", "err", err)
+			return err
+		}
+
+		// Get denom dynamically
+		hz := hostZone
+		balance := queryRes.Balances.AmountOf(hz.HostDenom)
+
+		// Set delegation account balance to ICQ result
+		hz, found := k.GetHostZone(ctx, hostZone.ChainId)
+		if found {
+			k.Logger(ctx).Error("invalid chain id, zone for \"%s\" already registered", hostZone.ChainId)
+		}
+
+		da := hz.DelegationAccount
+		da.Balance = balance.Int64()
+		hz.DelegationAccount = da
+		k.SetHostZone(ctx, hz)
+
+		ctx.EventManager().EmitEvents(sdk.Events{
+			sdk.NewEvent(
+				sdk.EventTypeMessage,
+				sdk.NewAttribute("totalUndelegatedBalance", balance.String()),
+			),
+		})
+
+		return nil
+	}
+	// 1. query delegation account undelegated balances using icq
+	// 2. write the result to hostZone.delegationAccount.Balance
+	k.InterchainQueryKeeper.QueryBalances(ctx, hostZone, cb, delegationAccount.Address)
+	return nil
+}
+
+// Update the redemption rate using values of delegatedBalances, balances and stAsset supply
+// TODO(TEST-97) add safety logic that checks balance, delegatedBalance and stAsset supply's block_height_updated are all equal
+func (k Keeper) UpdateExchangeRate(ctx sdk.Context, index int64, hostZone types.HostZone) error {
+	_ = ctx
+
+	// Assets: native asset balances on delegation account + staked
+	delegatedBalance := hostZone.GetDelegationAccount().DelegatedBalance
+	unDelegatedBalance := hostZone.GetDelegationAccount().Balance
+	assetBalance := delegatedBalance + unDelegatedBalance
+
+	// Claims: stAsset supply
+	stAssetSupply := k.bankKeeper.GetSupply(ctx, hostZone.IBCDenom)
+
+	// Sanity & safety check: if either num or denom are 0, do NOT update the exchange rate
+	if assetBalance == int64(0) || stAssetSupply.Amount.Int64() == int64(0) {
+		ctx.EventManager().EmitEvents(sdk.Events{
+			sdk.NewEvent(
+				sdk.EventTypeMessage,
+				sdk.NewAttribute("delegatedBalance", strconv.FormatInt(delegatedBalance, 10)),
+				sdk.NewAttribute("unDelegatedBalance", strconv.FormatInt(unDelegatedBalance, 10)),
+				sdk.NewAttribute("stAssetBalance", stAssetSupply.String()),
+			),
+		})
+		return errors.New("Exchange rate calculation error: ")
+	} else {
+		// ExchRate = Assets / Claims
```

Comment:
Should we refer to "redemption rate" from now on?

Exchange rate seems slightly confusing because you don't know which asset is the base. E.g. with 10 assets and 5 stAssets, the exchange rate is 2 and 0.5. If we refer to a redemption rate, it's clear that it's 2 (redeem 1 stAsset for 2 assets)

Reply:
Agreed